packages/xforms-engine/src/parse/xpath/semantic-analysis.ts

```diff
@@ -0,0 +1,290 @@
+	readonly text: LocalNameLiteral<LocalName>;
+}
+
+// prettier-ignore
+type LocalNamedFunctionCallLiteral<
+	LocalName extends string
+> = `${LocalNameLiteral<LocalName>}(${string})`;
+
+interface LocalNamedFunctionCallNode<LocalName extends string> extends FunctionCallNode {
+	readonly children: readonly [
+		name: LocalNamedFunctionNameNode<LocalName>,
+		...arguments: ArgumentNode[],
+	];
+	readonly text: LocalNamedFunctionCallLiteral<LocalName>;
+}
+
+const isCallToLocalNamedFunction = <LocalName extends string>(
+	syntaxNode: FunctionCallNode,
+	localName: LocalName
+): syntaxNode is LocalNamedFunctionCallNode<LocalName> => {
+	const [functionNameNode] = syntaxNode.children;
+	const [localNameNode] = collectTypedNodes(['local_part', 'unprefixed_name'], functionNameNode);
+
+	return localNameNode?.text === localName;
+};
+
+type ArgumentTypes = readonly AnySyntaxType[];
+
+const hasCallSignature = (syntaxNode: FunctionCallNode, expected: ArgumentTypes[]): boolean => {
+	const [, ...argumentNodes] = syntaxNode.children;
+
+	if (argumentNodes.length === 0 && expected.length === 0) {
+		return true;
+	}
+
+	if (argumentNodes.length > expected.length) {
+		return false;
+	}
+
+	return expected.every((types, i) => {
+		const argumentNode = argumentNodes[i];
+
+		if (argumentNode == null) {
+			return false;
+		}
+
+		const [firstMatch] = collectTypedNodes(types, argumentNode);
+
+		return firstMatch != null && isCompleteSubExpression(argumentNode, firstMatch);
+	});
+};
+
+const isTranslationFunctionCall = (syntaxNode: FunctionCallNode): boolean => {
+	return (
+		isCallToLocalNamedFunction(syntaxNode, 'itext') &&
+		hasCallSignature(syntaxNode, [
+			['string_literal', 'absolute_location_path', 'relative_location_path'],
+		])
+	);
+};
+
+export type TranslationExpression = LocalNamedFunctionCallLiteral<'itext'>;
+
+/**
+ * Determines if an arbitrary XPath expression is (in whole) a translation
+ * expression (i.e. a call to `jr:itext`).
+ */
+export const isTranslationExpression = (
+	expression: string
+): expression is TranslationExpression => {
+	const { rootNode } = expressionParser.parse(expression);
+	const functionCallNode = findTypedPrincipalExpressionNode(['function_call'], rootNode);
+
+	if (functionCallNode == null) {
+		return false;
+	}
+
+	return isTranslationFunctionCall(functionCallNode);
+};
+
+const isCurrentFunctionCall = (syntaxNode: FunctionCallNode): boolean => {
+	return isCallToLocalNamedFunction(syntaxNode, 'current') && hasCallSignature(syntaxNode, []);
+};
+
+/**
+ * Predicate to determine if a FilterPathExpr (as currently produced by
+ * `tree-sitter-xpath`) is one of:
+ *
+ * - `current()`
+ * - `current()/...` (where `...` represents additional steps)
+ * - `current()//...` (^)
+ *
+ * @todo XPath grammar technically also allows for `current()[some-predicate]`,
+ * and our `tree-sitter-xpath` grammar/parser also allow for this. But
+ * `@getodk/xpath` types do not currently acknowledge this possibility.
```

Comment:
Yes, it's possible in a form, it's possible anywhere. It's part of the XPath grammar: [FilterExpr](https://www.w3.org/TR/1999/REC-xpath-19991116/#NT-FilterExpr). The grammar itself is overly permissive—it allows Predicates on things you can't predicate, such as numbers and string literals—but the inferrable intent is specifically to allow Predicates on node-set returning FunctionCalls and on parenthesized Exprs (implicitly those producing node-sets as well).

I'm curious what you mean by:

> The expression would need to appear in a place where it's bound to a repeat because only repeats can be filtered in this way. It would have to be in a relevance or read-only expression.

Are you saying that we should add special conditions to enforce this, restricting this aspect of XPath grammar to these specific areas of XForms functionality? Or just that expressions like that would only be produced (e.g. by XLSForm/pyxform) in those cases.

Reply:
I just mean that it would only be actually useful/desirable to filter a nodeset with more than one node, I think. Basically the todo feels relatively low priority unless there's a use case I'm missing.

Reply:
It's mostly acknowledging that this is a gap in `@getodk/xpath`, both at the type level for `SyntaxNode` variants and (consequently) at the implementation of that aspect of the spec. Agree it's low priority! I don't want to lose the context completely, as it's one of very few areas `xpath` deviates from the spec.